physics/drag_suite.F90

```diff
@@ -1257,44 +1068,57 @@ subroutine drag_suite_run(                                           &
 !
 !  calculate - (g)*d(tau)/d(pressure) and deceleration terms dtaux, dtauy
 !
+!  First, set taup (momentum flux) at model top equal to that of the layer
+!  interface just below the top, i.e., taup(km)
+!  The idea is to allow the momentum flux to go out the 'top'.  This
+!  ensures there is no GWD force at the top layer.
+!
+         taup(i,km+1) = taup(i,km)
          do k = kts,km
-            taud_ls(i,k) = 1. * (taup(i,k+1) - taup(i,k)) * csg / del(i,k)
-            taud_bl(i,k) = 1. * (taufb(i,k+1) - taufb(i,k)) * csg / del(i,k)
+            taud_ms(i,k) = (taup(i,k+1) - taup(i,k)) * csg / del(i,k)
+            taud_bl(i,k) = (taufb(i,k+1) - taufb(i,k)) * csg / del(i,k)
          enddo
 !
-!  limit de-acceleration (momentum deposition ) at top to 1/2 value
-!  the idea is some stuff must go out the 'top'
-         do klcap = lcap,km
-            taud_ls(i,klcap) = taud_ls(i,klcap) * factop
-            taud_bl(i,klcap) = taud_bl(i,klcap) * factop
-         enddo
 !
-!  if the gravity wave drag would force a critical line
-!  in the lower ksmm1 layers during the next deltim timestep,
-!  then only apply drag until that critical line is reached.
+!  if the gravity wave drag + blocking would force a critical line
+!  in the layers below 'sigma' pressure level = rlolev during the next deltim
+!  timestep, then only apply drag until that critical line is reached, i.e.,
+!  reduce drag to limit resulting wind components to zero
+!  Note: 'sigma' pressure = ak + pref*bk in units of Pa
 !
          do k = kts,kpblmax-1
-            if (k .le. kbl(i)) then
-               if ((taud_ls(i,k)+taud_bl(i,k)).ne.0.)                   &
+            if (ak(k)+pref*bk(k).ge.rlolev) then
```

Comment:
First of all, I believe you can get the level of "ak(k)+pref*bk(k).ge.rlolev" for all horizontal loops. Since ak, pref, bk and rlolev are the same for all horizontal loops, you can calculate the index of the level outside of i=its, im loop. As a result, you can use "do k = kts, the index of the level" instead of "do k = kts, kpblmax-1"
Secondly, You used the fixed vertical levels above ground with "ak(k)+pref*bk(k).ge.rlolev".  Maybe it is better to use height above ground instead of fixed vertical level?

Reply:
Thank you for these suggestions.  I moved the "_ak(k)+prefbk(k).ge.rlolev_" calculation outside of the "_i=its,im_" loop, and use it to calculate a new integer variable "_k_rlolev_", which is the horizontally-independent vertical level index closest to the sigma-pressure level _rlolev_ (usually 500hPa).  This k-value is now the upper limit of the _k_ loop.
Regarding your second suggestion, after conferring with S. Moorthi, we decided to use a fixed vertical level instead of height above the ground.  This way, we are able to apply the "_defac_" drag limiter at all geographic locations, including the Himalayas, for example, where the surface pressure is less than 500hPa.

Reply:
The only concern for your current method is that not all host models have ak and bk.

Reply:
Yes, you need to move the level test back into the i=its,im loop. You can use the following method 1:
       (phil(i,k)/g < 5500.)  ! 5500m is roughly equivalent to 500 hPa height; phil is geopotential height at model layer
                or method2
       (prsl(i,k)/psurface)*1.e5 < 0.5e5 ! over low terrain area, the level is similar to that from method 1, over extremely high terrain, for example, the height of the terrain is near 600 hPa (~4000 m), you can get the height level at 300 hPa (~9000 m), the thickness is still around 5000m. 
So both methods seem ok.

Reply:
You can use prsi(k=1) as the surface air pressure, then use (prsi(i,k)/prsi(i,1)). Surface air pressure is available in the CCPP meta definition.